examples/merge-sort.html

```diff
@@ -0,0 +1,155 @@
+      <label for='number'>Array size:</label>
+      <input id='number' type='number' value='1000000' />
+    </p>
+    <p>
+      <button id="button1">Sort randomized array</button>
+    </p>
+    <p>
+      <button id="button2">Sort randomized array (GC Optimized)</button>
+    </p>
+
+    <div id='container'>
+    </div>
+
+    <script>
+      const container = document.querySelector("#container");
+      const numberEl = document.querySelector("#number");
+      let randomNumbers = generateRandomArray();
+
+      numberEl.addEventListener("change", () => {
+        randomNumbers = generateRandomArray()
+      });
+      document.querySelector("#button1")
+        .addEventListener("click", () => runSort(mergeSort), false);
+      document.querySelector("#button2")
+        .addEventListener("click", () => runSort(mergeSortOptimized), false);
+
+      function runSort(fn) {
+        // Measure the timing with the User Timings API.
+        const name = fn.name;
+        performance.mark(`${name}-start`);
+
+        // Run the sort algorithm.
+        const result = fn(randomNumbers.slice());
+
+        // Get the timing information.
+        performance.mark(`${name}-end`);
+        performance.measure(name, `${name}-start`, `${name}-end`);
+        const duration = performance.getEntriesByName(name)[0].duration.toFixed(2);
+        performance.clearMarks();
+        performance.clearMeasures();
+
+        // Update the UI
+        container.innerHTML = `
+          <p>Unsorted array: ${randomNumbers.slice(0, 50)}</p>
+          <p>Sorted array: ${result.slice(0, 50)}</p>
+          <p>"${name}" duration: ${duration}ms</p>
+        `;
+      }
+
+      /**
+       * This is a typical merge sort operation. It's terse and simple, but does require
+       * the browser to allocate quite a few small arrays.
+       */
+      function mergeSort(A) {
+        let result;
+        if (A.length < 2) {
+          result = A;
+        } else {
+          const mid = Math.floor(A.length / 2);
+          const left = mergeSort(A.slice(0, mid));
+          const right = mergeSort(A.slice(mid));
+
+          result = merge(left, right);
+        }
+        return result;
+      }
+
+      function merge(A, B) {
+        const result = [];
+        while (A.length > 0 && B.length > 0) {
+          result.push(A[0] < B[0] ? A.shift() : B.shift());
+        }
+        return result.concat(A.length ? A : B);
+      }
+
+      /**
+       * This optimized merge sort pre-allocates another array of the same size to hold
+       * intermediate sorting values. This has the benefit of not requiring additional
+       * allocations while performing the merge. There are no small intermediate arrays,
+       * and expensive operations such as concatenation are not needed. In addition,
+       * there is no garbage collection pressure of cleaning up lots of small arrays.
+       */
+      function mergeSortOptimized(A, B = Array(A.length), p = 0, r = A.length - 1) {
```

Comment:
I thought this needed to be `new Array(A.length)`, but it looks like I was wrong.